File: 443. String Compression/solution.py
class Solution(object):
    def compress(self, chars):
        """
        :type chars: List[str]
        :rtype: int
        """
        # maintain two helper arrays
        char_arr = []
        count_arr = []
        for char in chars:
            if char_arr and char_arr[-1] == char:
                    count_arr[-1] += 1
            else:
                char_arr.append(char)
                count_arr.append(1)
        
        # crate compressed array
        sol = []
        for count, char in zip(count_arr, char_arr):
            sol.append(char)
            if count > 1:
                for c in str(count):
                    sol.append(c)
        
        # check if compression useful
        if len(sol) <= len(chars):
            i = 0
            while i < len(sol):
                chars[i] = sol[i]
                i += 1
                
            while len(chars) > i:
                chars.pop()
                
        return len(chars)

File: 443. String Compression/solution.java
class Solution {
    public int compress(char[] chars) {
        int start = 0, end = 0, writeHead = 0;
        
        while (end < chars.length) {
            if (chars[start] == chars[end]) {
                end++;
            } else {
                chars[writeHead] = chars[start];
                writeHead++;
                if (end - start > 1) {
                    String count = Integer.toString(end - start);
                    for (int i = 0; i < count.length(); i++) {
                        chars[writeHead++] = count.charAt(i);
                    }
                }
                start = end;
                end++;
            }
        }
        chars[writeHead] = chars[start];
        writeHead++;
        if (end - start > 1) {
            String count = Integer.toString(end - start);
            for (int i = 0; i < count.length(); i++) {
                chars[writeHead++] = count.charAt(i);
            }
        }
        return writeHead;
    }
}
